CS0051 code fix

baks opened the pull request:
Fixes #321 

Code fix for CS0051 and also initial structure for other errors related with inconsistent accessibility.

InconsistentAccesibilityProviders should return InconsistentAccessibilityInfo with TypeSyntax, new accessibility modifiers and code action message. With that information code fix is able to search type symbol, locate proper document in solution and change accessibility modifiers for that type.

@giggio, let  me know what do you think

giggio commented on src/CSharp/CodeCracker/Design/InconsistentAccessibility/InconsistentAccessibilityCodeFixProvider.cs:
```diff
@@ -0,0 +1,159 @@
+                        context.RegisterCodeFix(CodeAction.Create(info.CodeActionMessage, c => ChangeTypeAccessibilityInDocument(document, info.NewAccessibilityModifiers, typeSymbol.Locations[0].SourceSpan, c)), diagnostic);
+                    }
+                }
+            }
```
`RegisterCodeFixesAsync` is doing too much work. See #347 where we are already changing older analyzers to perform better. It seems there is a way to delegate all this work to the Lambda in `CodeAction.Create`. Please update the provider to match that pattern, unless we really need to do some work on this method.

baks replied:
Ok, I'll change that

baks replied:
`RegisterCodeFixesAsync` checks whether type declaration exists in solution.
I think we can't move this to `CodeAction.Create` because if the type declaration is only in the metadata we shouldn't propose code fix at all.

Actually we can move part that finds Document and in `RegisterCodeFixesAsync` only see wheter any `Location` is in source. Agree?

giggio commented:
@baks The type has to exist on the solution, otherwise you couldn't even be referencing it, right? For example:

File Person.cs in your solution:
````csharp
class Person
{
    void WalkTheDog(Dog dog)
    {
    }
}
````

File Dog.cs outside of your solution, coming to you as a reference from a dll:
````csharp
class Dog
{
}
````

This will actually raise `CS0122`, not `CS0051`.

This scenario would work even on the same solution, with files being on different projects.

I guess `CS0051` is only raised when a method can reference the type, but not expose it outside as a parameter. If the type cannot be referenced, then we get `CS0122`.

BTW, if you can rebase or merge the changes from master it would be great. I updated some testing infrastructure.

baks commented:
Ok, I'll merge with master.

@giggio, If you create class like that in other assembly:
```csharp
public class Dog
{
    protected class Walker
    {
    }
}
```
and you use it in your solution:
```csharp
public class EnhancedDog
{
    public EnhancedDog(Walker walker)
    {
    }
}
```
This will raise `CS0051`.
Or if you use `InternalsVisibleToAssembly()` in example that you provided it should also raise `CS0051` (I'm not sure about this on 100% percent, but I'll check this when I get home).

giggio commented:
Hm... you are right!
We do need that check, then. If you can move all the other code to the other methods, away from `RegisterCodeFixesAsync` method, it would be fine.

giggio commented:
BTW, another possible code fix would be to make the `EnhancedDog` constructor `protected`, right? Let's finish this one, then we can think about it later.

baks commented:
@giggio , exactly :+1: 

This is the same as other fix I proposed in #321

baks commented:
@giggio I merged with master and included PR comments about `RegisterCodeFixesAsync`

giggio commented on src/CSharp/CodeCracker/Design/InconsistentAccessibility/InconsistentAccessibilityCodeFixProvider.cs:
```diff
@@ -0,0 +1,169 @@
+
+            if(typeSymbol != null)
+            {
+                var locationInSource = typeSymbol.Locations.FirstOrDefault(location => location.IsInSource);
```
You may find more than one location for a type if the type is declared `partial`. I suggest you return a `Task<Location[]>` or something similar.

baks replied:
That's right. I'll also add test for `partial` declaration

giggio commented on src/CSharp/CodeCracker/Design/InconsistentAccessibility/InconsistentAccessibilityCodeFixProvider.cs:
```diff
@@ -0,0 +1,169 @@
+            var result = document;
+
+            var declaration = syntaxRoot.FindNode(typeLocation.SourceSpan) as MemberDeclarationSyntax;
+            if (declaration != null)
```
When we get to the code fix provider we have to be sure of what syntax we are dealing with. There should not be any situation where `SyntaxRoot.FindNode(typeLocation.SourceSpan)` is not a `MemberDeclarationSyntax`.
I suggest you change this to `(MemberDeclarationSyntax)SyntaxRoot.FindNode(typeLocation.SourceSpan)` and remove the null check.

baks replied:
:+1:

giggio commented on src/CSharp/CodeCracker/Design/InconsistentAccessibility/InconsistentAccessibilityInMethodParameter.cs:
```diff
@@ -0,0 +1,107 @@
+            var parameterTypeNameStart = compilerErrorMessage.IndexOf(InconsistentAccessibilityCodeFixProvider.InconsistentAccessibilityInMethodParameterCompilerErrorNumber, StringComparison.Ordinal) + parameterTypeStartShift;
+            var parameterTypeNameLength = compilerErrorMessage.IndexOf('\'', parameterTypeNameStart) - parameterTypeNameStart;
+            return compilerErrorMessage.Substring(parameterTypeNameStart, parameterTypeNameLength);
+        }
```
You should not extract the parameter this way. We will have people that have different cultures in VS and the message will appear in different positions.
Try to change it to something like this:

````csharp
System.Text.RegularExpressions.Regex.Match(
    Microsoft.CodeAnalysis.CSharp.CSharpDiagnosticFormatter.Instance.Format(
        diagnostic, System.Globalization.CultureInfo.InvariantCulture),
        diagnostic.Descriptor.MessageFormat.ToString(System.Globalization.CultureInfo.InvariantCulture).Replace("{0}", "(.*)").Replace("{1}", "(.*)")
).Groups[1].Value
````

baks replied:
I'll change that to way that you proposed.

baks commented:
@giggio , I changed code fix provider to handle partial classess and also changed extracting parameter type to be culture agnostic with a way that you provided

giggio commented:
@baks The code is fine, but I found an inconsistent behavior on VS. I filed a bug here on the Roslyn repo: https://github.com/dotnet/roslyn/issues/3496
Because of that I think I will revert to a specific message until this is fixed. Tests pass but no code fix is offered at VS.

giggio commented:
@baks See 9674315 where I updated the method.
